lib/trento_web/controllers/tags_controller.ex

```diff
@@ -0,0 +1,84 @@
+defmodule TrentoWeb.TagsController do
+  use TrentoWeb, :controller
+
+  alias Trento.Tags
+
+  @spec add_tag(Plug.Conn.t(), map) :: Plug.Conn.t()
+  def add_tag(
+        conn,
+        %{
+          "id" => id,
+          "value" => value,
+          "resource" => resource
+        }
+      ) do
+    with {:ok, resource_type} <- resource_type_from_path(resource),
+         :ok <- guard_against_empty_tag(value),
+         {:ok, _} <- Tags.create_tag(value, id, resource_type) do
+      conn
+      |> put_status(:created)
+      |> json(%{value: value})
+    else
+      {:error, :unrecognized_resource} ->
+        unrecognized_resource_error(conn, resource)
+
+      {:error, :empty_tag_detected} ->
+        empty_tag_error(conn)
+
+      {:error, _} ->
+        creation_error(conn)
+    end
+  end
+
+  @spec remove_tag(Plug.Conn.t(), map) :: Plug.Conn.t()
+  def remove_tag(conn, %{
+        "id" => resource_id,
+        "value" => value
+      }) do
+    case Tags.delete_tag(value, resource_id) do
+      :ok ->
+        send_resp(conn, :no_content, "")
+
+      :not_found ->
+        send_resp(conn, :not_found, "")
+    end
+  end
+
+  defp resource_type_from_path("hosts"), do: {:ok, :host}
+  defp resource_type_from_path("clusters"), do: {:ok, :cluster}
+  defp resource_type_from_path("sap_systems"), do: {:ok, :sap_system}
+  defp resource_type_from_path("databases"), do: {:ok, :database}
+  defp resource_type_from_path(_), do: {:error, :unrecognized_resource}
+
+  defp empty?(<<" "::binary, rest::binary>>), do: empty?(rest)
+  defp empty?(""), do: true
+  defp empty?(_string), do: false
+
```

Comment:
not sure what this does? couldn't you check the length of the string?

Reply:
this guy checks for empty and blank strings :smile: